firedrake/functionspaceimpl.py

```diff
@@ -0,0 +1,736 @@
+        if type(element) is ufl.MixedElement:
+            raise ValueError("Can't create FunctionSpace for MixedElement")
+        fiat_element = create_element(element, vector_is_mixed=False)
+        sdata = get_shared_data(mesh, fiat_element)
+        # The function space shape is the number of dofs per node,
+        # hence it is not always the value_shape.  Vector and Tensor
+        # element modifiers *must* live on the outside!
+        if type(element) is ufl.TensorElement:
+            # UFL enforces value_shape of the subelement to be empty
+            # on a TensorElement.
+            self.shape = element.value_shape()
+        elif type(element) is ufl.VectorElement:
+            # First dimension of the value_shape is the VectorElement
+            # shape.
+            self.shape = element.value_shape()[:1]
+        else:
+            self.shape = ()
+        self._ufl_element = element
+        self._shared_data = sdata
+        self._mesh = mesh
+
+        self.rank = len(self.shape)
+        """The rank of this :class:`FunctionSpace`.  Spaces where the
+        element is scalar-valued (or intrinsically vector-valued) have
+        rank zero.  Spaces built on :class:`~ufl.classes.VectorElement` or
+        :class:`~ufl.classes.TensorElement` instances have rank equivalent to
+        the number of components of their
+        :meth:`~ufl.classes.FiniteElementBase.value_shape`."""
+        self.dim = numpy.prod(self.shape, dtype=int)
+        """The total number of degrees of freedom at each function
+        space node."""
+        self.name = name
+        """The (optional) descriptive name for this space."""
+        self.node_set = sdata.node_set
+        """A :class:`pyop2.Set` representing the function space nodes."""
+        self.dof_dset = op2.DataSet(self.node_set, self.shape or 1,
+                                    name="%s_nodes_dset" % self.name)
+        """A :class:`pyop2.DataSet` representing the function space
+        degrees of freedom."""
+        self.fiat_element = fiat_element
+        self.extruded = sdata.extruded
+        self.offset = sdata.offset
+        self.bt_masks = sdata.bt_masks
+
+    # These properties are overridden in ProxyFunctionSpaces, but are
+    # provided by FunctionSpace so that we don't have to special case.
+    index = None
+    """The position of this space in its parent
+    :class:`MixedFunctionSpace`, or ``None``."""
+
+    parent = None
+    """The parent space if this space was extracted from one, or ``None``."""
+
+    component = None
+    """The component of this space in its parent VectorElement space, or
+    ``None``."""
+
+    def __eq__(self, other):
+        try:
+            # FIXME: Think harder about equality
+            return self.mesh() is other.mesh() and \
+                self.dof_dset is other.dof_dset and \
+                self.ufl_element() == other.ufl_element()
+        except AttributeError:
+            return False
```

Comment:
Is this the correct equality comparison?  I think if the meshes match, and the ufl element matches, the dof_dset must also match, but I was belt-and-braces here.